src/cljs/ataru/hakija/hakija_readonly.cljs

```diff
@@ -18,40 +18,60 @@
                                                                        textual-field-value
                                                                        scroll-to-anchor
                                                                        question-group-answer?
-                                                                       answers->read-only-format]]
+                                                                       answers->read-only-format
+                                                                       value-or-koodi-uri->label
+                                                                       group-spacer]]
             [taoensso.timbre :refer-macros [spy debug]]))
 
 (defn- multiple-choice-with-koodisto [field-descriptor]
   (and (= (:fieldType field-descriptor) "multipleChoice")
        (contains? field-descriptor :koodisto-source)))
 
-(defn text [field-descriptor application lang]
+(defn- multi-values->:li
+  [field-descriptor lang values]
+  (map-indexed (fn [idx value-or-koodi-uri]
+                 (let [value (str (if (map? value-or-koodi-uri)
+                                    (:value value-or-koodi-uri)
+                                    value-or-koodi-uri))]
+                   ^{:key (str "value-" idx)}
+                   [:li (value-or-koodi-uri->label field-descriptor lang value)]))
+               values))
+
+(defn text [field-descriptor application lang question-group-index]
   (let [answer            ((answer-key field-descriptor) (:answers application))
-        multi-values->:li (partial map-indexed (fn [idx value]
-                                                 ^{:key (str "value-" idx)}
-                                                 [:li (str (if (map? value)
-                                                             (:value value)
-                                                             value))]))]
+        values (if question-group-index
+                 (-> answer :values (nth question-group-index))
+                 (:values answer))]
     [:div.application__form-field
      [:label.application__form-field-label
       (str (-> field-descriptor :label lang) (required-hint field-descriptor))]
      (if (:cannot-view answer)
        [:div "***********"]
-       [:div
-        (cond (and (vector? (:values answer))
-                   (every? vector? (:values answer)))
-              (into [:ul.application__form-field-list] (map multi-values->:li (:values answer)))
-
-              (and (vector? (:values answer))
-                   (some (comp not vector?) (:values answer)))
-              (into [:ul.application__form-field-list] (multi-values->:li (:values answer)))
+       [:div.application__readonly-text
+        (cond (and (vector? values)
+                   (every? vector? values)
```

Comment:
Should a `(not (empty? values))` check be in the `and` clause here too? Invoking

```clojure
(def values [])
(and (vector? values)
     (every? vector? values))
```

will return `true`. Should also note that there are a lot of other places to be fixed regarding this.

Reply:
Maybe we should have an util function to actually check this and then use that function everywhere?

Reply:
In that case will (map) over empty so no problem.